cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2758,6 +2758,318 @@ struct GroupedOneFactory {
+    auto other = checked_cast<GroupedListImpl*>(&raw_other);
+    auto other_raw_values = other->values_.mutable_data();
+    uint32_t* other_raw_groups = other->groups_.mutable_data();
+    auto g = group_id_mapping.GetValues<uint32_t>(1);
+
+    for (uint32_t other_g = 0; static_cast<int64_t>(other_g) < other->num_args_;
+         ++other_g) {
+      if (bit_util::GetBit(other->values_bitmap_.data(), other_g)) {
+        RETURN_NOT_OK(values_.Append(GetSet::Get(other_raw_values, other_g)));
+        RETURN_NOT_OK(values_bitmap_.Append(true));
+      } else {
+        RETURN_NOT_OK(values_.Append(static_cast<CType>(0)));
+        RETURN_NOT_OK(values_bitmap_.Append(false));
+      }
+      RETURN_NOT_OK(groups_.Append(g[other_raw_groups[other_g]]));
+      ++num_args_;
+    }
+    return Status::OK();
+  }
+
+  Result<Datum> Finalize() override {
+    ARROW_ASSIGN_OR_RAISE(auto values_buffer, values_.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto groups_buffer, groups_.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto null_bitmap_buffer, values_bitmap_.Finish());
+
+    auto groups = UInt32Array(num_args_, std::move(groups_buffer));
+    ARROW_ASSIGN_OR_RAISE(
+        auto groupings,
+        Grouper::MakeGroupings(groups, static_cast<uint32_t>(num_groups_), ctx_));
+
+    auto values_array_data = ArrayData::Make(
+        out_type_, num_args_, {std::move(null_bitmap_buffer), std::move(values_buffer)});
+    auto values = MakeArray(values_array_data);
+    return Grouper::ApplyGroupings(*groupings, *values);
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return list(out_type_); }
+
+  ExecContext* ctx_;
+  int64_t num_groups_, num_args_ = 0;
+  TypedBufferBuilder<CType> values_;
+  TypedBufferBuilder<uint32_t> groups_;
+  TypedBufferBuilder<bool> values_bitmap_;
+  std::shared_ptr<DataType> out_type_;
+};
+
+template <typename Type>
+struct GroupedListImpl<Type, enable_if_t<is_base_binary_type<Type>::value ||
+                                         std::is_same<Type, FixedSizeBinaryType>::value>>
+    final : public GroupedAggregator {
+  using Allocator = arrow::stl::allocator<char>;
+  using StringType = std::basic_string<char, std::char_traits<char>, Allocator>;
+  using GetSet = GroupedValueTraits<Type>;
+
+  Status Init(ExecContext* ctx, const std::vector<ValueDescr>&,
+              const FunctionOptions* options) override {
+    ctx_ = ctx;
+    allocator_ = Allocator(ctx->memory_pool());
+    // out_type_ initialized by GroupedListInit
+    groups_ = TypedBufferBuilder<uint32_t>(ctx_->memory_pool());
+    values_bitmap_ = TypedBufferBuilder<bool>(ctx_->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return Status::OK();
+  }
+
+  Status Consume(const ExecBatch& batch) override {
+    return VisitGroupedValues<Type>(
+        batch,
+        [&](uint32_t group, util::string_view val) -> Status {
+          values_.emplace_back(val);
```

Comment:
Have simply used `emplace_back` as opposed to `emplace` as used here:
https://github.com/apache/arrow/blob/5216c2bb3e4f9298230e2e38f29ff9889b5a152d/cpp/src/arrow/compute/kernels/hash_aggregate.cc#L1853
Is this valid or will something like resizing and then using `emplace` will have to be used? (As `allocator_` is not used explicitly in `emplace_back`)

Reply:
allocator_ is not a parameter of emplace in the first place. emplace/emplace_back forward their arguments so really it's part of the std::basic_string constructor. And hence allocator_ should still work here.